controlplane/kubeadm/internal/filters.go

```diff
@@ -41,65 +40,73 @@ import (
 // - mutated in-place (ex: NodeDrainTimeout)
 // - are not dictated by KCP (ex: ProviderID)
 // - are not relevant for the rollout decision (ex: failureDomain).
-func matchesMachineSpec(infraConfigs map[string]*unstructured.Unstructured, machineConfigs map[string]*bootstrapv1.KubeadmConfig, kcp *controlplanev1.KubeadmControlPlane, machine *clusterv1.Machine) (string, bool, error) {
-	mismatchReasons := []string{}
+func matchesMachineSpec(infraConfigs map[string]*unstructured.Unstructured, machineConfigs map[string]*bootstrapv1.KubeadmConfig, kcp *controlplanev1.KubeadmControlPlane, machine *clusterv1.Machine) (bool, []string, []string, error) {
+	logReasons := []string{}
+	conditionReasons := []string{}
 
 	if !collections.MatchesKubernetesVersion(kcp.Spec.Version)(machine) {
 		machineVersion := ""
 		if machine != nil && machine.Spec.Version != nil {
 			machineVersion = *machine.Spec.Version
 		}
-		mismatchReasons = append(mismatchReasons, fmt.Sprintf("Machine version %q is not equal to KCP version %q", machineVersion, kcp.Spec.Version))
+		logReasons = append(logReasons, fmt.Sprintf("Machine version %q is not equal to KCP version %q", machineVersion, kcp.Spec.Version))
+		conditionReasons = append(conditionReasons, "spec.version is not up to date")
 	}
 
 	reason, matches, err := matchesKubeadmBootstrapConfig(machineConfigs, kcp, machine)
 	if err != nil {
-		return "", false, errors.Wrapf(err, "failed to match Machine spec")
+		return false, nil, nil, errors.Wrapf(err, "failed to match Machine spec")
 	}
 	if !matches {
-		mismatchReasons = append(mismatchReasons, reason)
+		logReasons = append(logReasons, reason)
+		conditionReasons = append(conditionReasons, "spec.bootstrap is not up to date")
```

Comment:
```suggestion
		conditionReasons = append(conditionReasons, "KubeadmConfig is not up to date")
```
maybe, seems more concrete and more correct (given that we also compare the ClusterConfiguration on the Machine object)